iXBRLViewerPlugin/viewer/src/js/outline.js

```diff
@@ -0,0 +1,168 @@
+// limitations under the License.
+
+import $ from 'jquery'
+
+// DocumentOutline chooses a fact for each presentation group (ELR) that
+// represents the start of that ELR.  This is done by deciding which ELRs each
+// fact participates in (see factInGroup()) and then finding the longest
+// continuous run of facts in document order that participate in each ELR.
+export function DocumentOutline(report) {
+    this._report = report;
+    var facts = report.facts().sort((a, b) => a.ixNode.docOrderindex - b.ixNode.docOrderindex);
+    var runLength = {};
+    var runStart = {};
+    var longestRun = {};
+    var longestRunStart = {};
+    this._buildDimensionMap();
+    const elrs = report.relationshipGroups("pres");
+    for (const f of facts) {
+        if (f.isHidden()) {
+            continue;
+        }
+        for (const elr of elrs) {
+            if (this.factInGroup(f, elr)) {
+                if (!(elr in runLength)) {
+                    // Start counting a run
+                    runLength[elr] = 0;
+                    runStart[elr] = f;
+                }
+                runLength[elr]++;
+            }
+            else if (elr in runLength) { 
+                // End of a run
+                if (!(elr in longestRun) || longestRun[elr] < runLength[elr]) {
+                    longestRun[elr] = runLength[elr];
+                    longestRunStart[elr] = runStart[elr];
+                }
+                delete runLength[elr];
+                delete runStart[elr];
+            }
+        }
+    }
+
+    // End of document, check if any current runs are the longest run for the
+    // ELR.
+    for (const elr in runLength) {
+        if (!(elr in longestRun) || longestRun[elr] < runLength[elr]) {
+            longestRun[elr] = runLength[elr];
+            longestRunStart[elr] = runStart[elr];
+        }
+    }
+
+    this.sections = longestRunStart;
+}
+
+// Returns true if a fact participates in the given presentation group.
+DocumentOutline.prototype.factInGroup = function (fact, elr) {
+    // Roots are abstract so no need to check for concepts with outgoing
+    // relationships only.
+    if (this._report.getParentRelationshipsInGroup(fact.conceptName(), "pres", elr).length == 0) {
+        return false;
+    }
+    const fd = fact.dimensions();
+    const dm = this.dimensionMap[elr];
+    // Check all dimensions specified in this ELR
+    for (const [dim, spec] of Object.entries(dm)) {
+        // If a fact has a dimension, it must be in the list of permitted
+        // members, otherwise, the default member must be allowed
+        if (spec.typed) {
+            if (!(dim in fd)) {
+                return false;
+            }
+        }
+        else if ((dim in fd) ? !(fd[dim] in spec.members) : !spec.allowDefault) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// Build a map of ELRs to dimensional information:
+//   { elr: 
+//      { dimensionQName: 
```

Comment:
small nit: probably a bit easier with the keys all on a newline and the brackets on the same line as the key thats holding the map i.e. something like:

```
{ 
    elr: {
         dimensionQname: {
                allowDefaults: true,
                         members: {
                                   memberQname: true
                         }
           }
}
```